server/src/main/java/io/crate/expression/scalar/timestamp/CurrentTimeFunction.java

```diff
@@ -21,38 +21,81 @@
-import io.crate.metadata.SystemClock;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
 import io.crate.metadata.functions.Signature;
 import io.crate.types.DataTypes;
 import io.crate.types.TimeTZ;
 
-import static io.crate.metadata.SystemClock.microsFromMidnightUTC;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Locale;
 
-public class CurrentTimeFunction extends TimeFunction<TimeTZ> {
+public class CurrentTimeFunction extends Scalar<TimeTZ, Integer> {
 
     public static final String NAME = "current_time";
+    private static final int MICRO_PRECISION = 6; // microseconds
 
     public static void register(ScalarFunctionModule module) {
         module.register(
             Signature.scalar(
                 NAME,
                 DataTypes.INTEGER.getTypeSignature(),
-                DataTypes.TIMETZ.getTypeSignature()),
-            (signature, args) -> new CurrentTimeFunction(signature));
+                DataTypes.TIMETZ.getTypeSignature()
+            ),
+            CurrentTimeFunction::new
+        );
         module.register(
             Signature.scalar(
                 NAME,
-                DataTypes.TIMETZ.getTypeSignature()),
-            (signature, args) -> new CurrentTimeFunction(signature));
+                DataTypes.TIMETZ.getTypeSignature()
+            ),
+            CurrentTimeFunction::new
+        );
     }
 
-    private CurrentTimeFunction(Signature signature) {
-        super(signature, SystemClock.MICRO_PRECISION, (i, p) -> new TimeTZ(microsFromMidnightUTC(i, p), 0));
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private CurrentTimeFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final TimeTZ evaluate(TransactionContext txnCtx, Input<Integer>... args) {
+        Integer precision = MICRO_PRECISION;
+        if (args.length == 1) {
+            precision = args[0].value();
+            if (precision == null) {
+                throw new IllegalArgumentException(String.format(
+                    Locale.ENGLISH,
+                    "NULL precision not supported for %s",
+                    signature.getName().name()));
+            }
+        }
+        if (precision < 0 || precision > MICRO_PRECISION) {
+            throw new IllegalArgumentException(String.format(
+                Locale.ENGLISH, "precision must be between [0..%d]", MICRO_PRECISION));
+        }
+
+        Instant i = txnCtx.currentInstant();
+        long dateTimeMicros = (i.getEpochSecond() * 1000_000_000L + i.getNano()) / 1000L;
+        long dateMicros = i.truncatedTo(ChronoUnit.DAYS).toEpochMilli() * 1000L;
+        long f = (long) Math.pow(10, MICRO_PRECISION - precision);
+        return new TimeTZ(((dateTimeMicros - dateMicros) / f) * f, 0);
```

Comment:
Would something like this works?

```suggestion
        Instant currentInstant = txnCtx.currentInstant();
        long microsSinceMidnight = ChronoUnit.MICROS.between(currentInstant.truncatedTo(ChronoUnit.DAYS), currentInstant);
        long factor = (long) Math.pow(10, MICRO_PRECISION - precision);
        return new TimeTZ((microsSinceMidnight / factor) * factor, 0);
```

This would make the conversion a bit more declarative and with that probably a bit easier to read.

Reply:
oh! yes, higher abstraction is better. For me it is more readable the way I wrote it, but that is subjective. If allowed I will go functional.